src/imp/libc/process/syscalls.rs

```diff
@@ -288,6 +292,51 @@ pub(crate) fn getrlimit(limit: Resource) -> Rlimit {
     }
 }
 
+#[cfg(not(any(target_os = "fuchsia", target_os = "redox", target_os = "wasi")))]
+#[inline]
+pub(crate) fn setrlimit(limit: Resource, new: Rlimit) -> io::Result<()> {
+    let rlim_cur = match new.current {
+        Some(r) => r.try_into().map_err(|_| io::Error::INVAL)?,
+        None => LIBC_RLIM_INFINITY as _,
+    };
+    let rlim_max = match new.maximum {
+        Some(r) => r.try_into().map_err(|_| io::Error::INVAL)?,
+        None => LIBC_RLIM_INFINITY as _,
+    };
+    let lim = libc_rlimit { rlim_cur, rlim_max };
+    unsafe { ret(libc_setrlimit(limit as _, as_ptr(&lim))) }
+}
+
+#[cfg(any(target_os = "android", target_os = "linux"))]
+#[inline]
+pub(crate) fn prlimit(pid: Option<Pid>, limit: Resource, new: Rlimit) -> io::Result<Rlimit> {
+    let lim = libc_rlimit {
+        rlim_cur: new.current.unwrap_or(LIBC_RLIM_INFINITY),
+        rlim_max: new.maximum.unwrap_or(LIBC_RLIM_INFINITY),
+    };
+    let mut result = MaybeUninit::<libc_rlimit>::uninit();
+    unsafe {
+        ret_infallible(libc_prlimit(
+            Pid::as_raw(pid),
+            limit as _,
+            as_ptr(&lim),
+            result.as_mut_ptr(),
+        ));
+        let result = result.assume_init();
+        let current = if result.rlim_cur == LIBC_RLIM_INFINITY {
+            None
+        } else {
+            result.rlim_cur.try_into().ok()
+        };
```

Comment:
I am not totally following the `try_into()` bits here around the edge/maximum cases.  May be worth factoring this into a helper function that has a bit of commentary?

(Here and elsewhere)

Reply:
Good idea, this was getting pretty cluttered. I've now factored out helper functions to do the conversions, and fixed those `.try_into().ok()` things, which we shouldn't ever do.